From: Gerald Schaefer <gerald.schaefer@de.ibm.com>
Subject: s390/pci: fix use after free in dma_init
Patch-mainline: v4.6-rc6
Git-commit: dba599091c191d209b1499511a524ad9657c0e5a
References: bnc#980892, LTC#141626

Description:  s390/pci: fix use after free in dma_init
Symptom:      Possible memory corruption.
Problem:      After a failure during registration of the dma_table (because
              of the function being in error state) we free its memory but
              don't reset the associated pointer to zero.
              When we then receive a notification from firmware (about the
              function being in error state) we'll try to walk and free the
              dma_table again.
Solution:     Fix this by resetting the dma_table pointer. In addition to
              that make sure that we free the iommu_bitmap when appropriate.
Reproduction: Hard to hit, attach/detach a PCI function under z/VM in a loop.

Upstream-Description:

              s390/pci: fix use after free in dma_init

              After a failure during registration of the dma_table (because of the
              function being in error state) we free its memory but don't reset the
              associated pointer to zero.

              When we then receive a notification from firmware (about the function
              being in error state) we'll try to walk and free the dma_table again.

              Fix this by resetting the dma_table pointer. In addition to that make
              sure that we free the iommu_bitmap when appropriate.

              Signed-off-by: Sebastian Ott <sebott@linux.vnet.ibm.com>
              Reviewed-by: Gerald Schaefer <gerald.schaefer@de.ibm.com>
              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Gerald Schaefer <gerald.schaefer@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/pci/pci_dma.c |   16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

--- a/arch/s390/pci/pci_dma.c
+++ b/arch/s390/pci/pci_dma.c
@@ -458,7 +458,7 @@ int zpci_dma_init_device(struct zpci_dev
 	zdev->dma_table = dma_alloc_cpu_table();
 	if (!zdev->dma_table) {
 		rc = -ENOMEM;
-		goto out_clean;
+		goto out;
 	}
 
 	/*
@@ -478,18 +478,22 @@ int zpci_dma_init_device(struct zpci_dev
 	zdev->iommu_bitmap = vzalloc(zdev->iommu_pages / 8);
 	if (!zdev->iommu_bitmap) {
 		rc = -ENOMEM;
-		goto out_reg;
+		goto free_dma_table;
 	}
 
 	rc = zpci_register_ioat(zdev, 0, zdev->start_dma, zdev->end_dma,
 				(u64) zdev->dma_table);
 	if (rc)
-		goto out_reg;
-	return 0;
+		goto free_bitmap;
 
-out_reg:
+	return 0;
+free_bitmap:
+	vfree(zdev->iommu_bitmap);
+	zdev->iommu_bitmap = NULL;
+free_dma_table:
 	dma_free_cpu_table(zdev->dma_table);
-out_clean:
+	zdev->dma_table = NULL;
+out:
 	return rc;
 }
 
